internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+		return nil, err
+	}
+	if _, err = attachIOCP(jobHandle, iocpHandle); err != nil {
+		windows.Close(jobHandle)
+		windows.Close(iocpHandle)
+		return nil, err
+	}
+	return &jobObject{jobHandle, iocpHandle}, nil
+}
+
+// close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *jobObject) close() error {
+	closeErr := false
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			closeErr = true
+		}
+		job.jobHandle = 0
+	}
+	if job.iocpHandle != 0 {
+		if err := windows.Close(job.iocpHandle); err != nil {
+			closeErr = true
+		}
+		job.iocpHandle = 0
+	}
+	if closeErr {
+		return errors.New("failed to close one or more handles")
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *jobObject) assign(p *JobProcess) error {
+	if p.Pid() == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, uint32(p.Pid()))
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *jobObject) terminate() error {
+	if job.jobHandle != 0 {
+		return windows.TerminateJobObject(job.jobHandle, 1)
+	}
+	return nil
+}
+
+func (job *jobObject) shutdown(ctx context.Context) error {
+	pids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	var (
+		terminate bool
+		signalErr bool
+	)
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
+	return nil
+}
+
+// Returns all of the process IDs in the job object.
+func (job *jobObject) pids() ([]uint32, error) {
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	var retLen uint32
+	err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		&retLen,
+	)
+
+	// If there are no processes in the job the next call to QueryInformation will just
+	// hang until a memory alloc error. Return here.
+	if err == nil {
+		// Return empty slice instead of nil to play well with the caller of this.
+		// I don't think we should return an error on there being no PIDs.
```

Comment:
I think this last sentence of the comment either shouldn't exist OR should say something more like "do not return an error if no processes are running inside the job"

Reply:
Agreed, forgot to reword this when making this